lib/dao.js

```diff
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.patchOrCreateWithWhere =
+DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) { return connectionPromise; }
+  assert(arguments.length >= 1, 'At least one argument is required');
+  if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
```

Comment:
IIUC, this is handling `upsertWithWhere(instanceData)`, which means the query will return all instances in the database, and the operation succeeds only if there is at most one model instance. Such usage seem a bit weird to me. Could you please explain where would LoopBack users want to call `upsertWithWhere` this way?